workflow/sync/rebalanced.go

```diff
@@ -0,0 +1,144 @@
+package sync
+
+import (
+	"math"
+	"time"
+
+	wfv1 "github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
+)
+
+// TODO: maintain cache, expire entries
+
+// RebalanceQueue implements the rebalance strategy for consumption in Semaphore
+type RebalanceQueue struct {
+	items             []*item           // standard item with an attached rebalance key
+	rebalanceKeyCache map[string]string // mapping from resource name (holder) to holder's rebalance key
+	semaphore         Semaphore         // reference to parent semaphore, so we can get holders
+}
+
+func NewRebalanceQueue() *RebalanceQueue {
+	return &RebalanceQueue{
+		items:             make([]*item, 0),
+		rebalanceKeyCache: make(map[string]string, 0),
+	}
+}
+
+// RebalanceQueue needs to reference semaphore to know current lock holders
+func (r *RebalanceQueue) setParentSemaphore(s Semaphore) {
+	r.semaphore = s
+}
+
+// determines the best ordering based on currently-outstanding keys
+func (r *RebalanceQueue) onRelease(key Key) {
+	// we need pending + holding counts for rebalance keys, and just holding counts
+	allRebalanceKeys := make(map[string]int, 0)
+	holderRebalanceKeys := make(map[string]int, 0)
+
+	for _, h := range r.semaphore.getCurrentHolders() {
+		rk := r.rebalanceKeyCache[h]
+		allRebalanceKeys[rk] += 1
+		if h != key {
+			holderRebalanceKeys[rk] += 1
+		}
+	}
+
+	for _, p := range r.items {
+		rk := r.rebalanceKeyCache[p.key]
+		allRebalanceKeys[rk] += 1
+	}
+
+	// partition r.queue into items that can / can't be scheduled
+	can := make([]*item, 0)
+	overflow := make(map[string]*item, 0)
+	cant := make([]*item, 0)
+
+	maxLocksPerUserFloor := math.Floor(float64(r.semaphore.getLimit()) / float64(len(allRebalanceKeys)))
+	// iterate through pending queue to determine what has changed
+	for _, p := range r.items {
+		if p.key != key {
+			rebalanceKey := r.rebalanceKeyCache[p.key]
+			// we want to ignore "key" since removal from queue happens after release in semaphore
+			if float64(holderRebalanceKeys[rebalanceKey]) < maxLocksPerUserFloor {
```

Comment:
Minor thing but why convert this to a float? Is it not better to cast maxLocksPerUserFloor into an int instead? 
I know we will probably never hit numbers large enough where comparison being imprecise is a worry, but probs nice to use ints regardless.

Reply:
@isubasinghe it's because `maxLocksPerUserFloor` is a float, since Floor returns float64. See line 55.

Reply:
Yes, I was more suggesting to go with approach on line 67 instead of casting to floats. It honestly is a very minor nit though, probs fine as it is, this is more me just being paranoid